Without the changes in this pull request, the following fails in Python 2
```python
x = np.zeros(1, dtype=[(u'test', int)])
x.sort(order=u'test')
```
with
```
Traceback (most recent call last):
  File "91.py", line 4, in <module>
    x.sort(order=u'test')
  File "/home/bilbo/anaconda2/lib/python2.7/site-packages/numpy/core/_internal.py", line 312, in _newnames
    raise ValueError("unsupported order value: %s" % (order,))
ValueError: unsupported order value: test
```

and the following

```python
x = np.core.records.fromarrays([[0], [0]], names=u'a,b', formats=u'i4,i4')
```

fails with
```
Traceback (most recent call last):
  File "91.py", line 6, in <module>
    x = np.core.records.fromarrays([[0], [0]], names=u'a,b', formats=u'i4,i4')
  File "/home/bilbo/anaconda2/lib/python2.7/site-packages/numpy/core/records.py", line 610, in fromarrays
    parsed = format_parser(formats, names, titles, aligned, byteorder)
  File "/home/bilbo/anaconda2/lib/python2.7/site-packages/numpy/core/records.py", line 147, in __init__
    self._setfieldnames(names, titles)
  File "/home/bilbo/anaconda2/lib/python2.7/site-packages/numpy/core/records.py", line 180, in _setfieldnames
    raise NameError("illegal input names %s" % repr(names))
NameError: illegal input names u'a,b'
```

Although the fix from pull request #10672 allowed for unicode field names in Python 2, the two above examples would still fail as the strings were explicitly required to be bytestrings before being passed to lower level code.

--- a/numpy/core/tests/test_multiarray.py
+++ b/numpy/core/tests/test_multiarray.py
@@ -4748,64 +4748,69 @@ class TestRecord(object):
         # Error raised when multiple fields have the same name
         assert_raises(ValueError, test_assign)
 
-    if sys.version_info[0] >= 3:
-        def test_bytes_fields(self):
-            # Bytes are not allowed in field names and not recognized in titles
-            # on Py3
-            assert_raises(TypeError, np.dtype, [(b'a', int)])
-            assert_raises(TypeError, np.dtype, [(('b', b'a'), int)])
+    @pytest.mark.skipif(sys.version_info[0] < 3, reason="Not Python 3")
+    def test_bytes_fields(self):
+        # Bytes are not allowed in field names and not recognized in titles
+        # on Py3
+        assert_raises(TypeError, np.dtype, [(b'a', int)])
+        assert_raises(TypeError, np.dtype, [(('b', b'a'), int)])
 
-            dt = np.dtype([((b'a', 'b'), int)])
-            assert_raises(TypeError, dt.__getitem__, b'a')
+        dt = np.dtype([((b'a', 'b'), int)])
+        assert_raises(TypeError, dt.__getitem__, b'a')
 
-            x = np.array([(1,), (2,), (3,)], dtype=dt)
-            assert_raises(IndexError, x.__getitem__, b'a')
+        x = np.array([(1,), (2,), (3,)], dtype=dt)
+        assert_raises(IndexError, x.__getitem__, b'a')
 
-            y = x[0]
-            assert_raises(IndexError, y.__getitem__, b'a')
+        y = x[0]
+        assert_raises(IndexError, y.__getitem__, b'a')
 
-        def test_multiple_field_name_unicode(self):
-            def test_assign_unicode():
-                dt = np.dtype([("\u20B9", "f8"),
-                               ("B", "f8"),
-                               ("\u20B9", "f8")])
+    @pytest.mark.skipif(sys.version_info[0] < 3, reason="Not Python 3")
+    def test_multiple_field_name_unicode(self):
+        def test_assign_unicode():
+            dt = np.dtype([("\u20B9", "f8"),
+                           ("B", "f8"),
+                           ("\u20B9", "f8")])
 
-            # Error raised when multiple fields have the same name(unicode included)
-            assert_raises(ValueError, test_assign_unicode)
+        # Error raised when multiple fields have the same name(unicode included)
+        assert_raises(ValueError, test_assign_unicode)
 
-    else:
-        def test_unicode_field_titles(self):
-            # Unicode field titles are added to field dict on Py2
-            title = u'b'
-            dt = np.dtype([((title, 'a'), int)])
-            dt[title]
-            dt['a']
-            x = np.array([(1,), (2,), (3,)], dtype=dt)
-            x[title]
-            x['a']
-            y = x[0]
-            y[title]
-            y['a']
+    @pytest.mark.skipif(sys.version_info[0] >= 3, reason="Not Python 2")
+    def test_unicode_field_titles(self):
+        # Unicode field titles are added to field dict on Py2
+        title = u'b'
+        dt = np.dtype([((title, 'a'), int)])
+        dt[title]
+        dt['a']
+        x = np.array([(1,), (2,), (3,)], dtype=dt)
+        x[title]
+        x['a']
+        y = x[0]
+        y[title]
+        y['a']
 
-        def test_unicode_field_names(self):
-            # Unicode field names are converted to ascii on Python 2:
-            encodable_name = u'b'
-            assert_equal(np.dtype([(encodable_name, int)]).names[0], b'b')
-            assert_equal(np.dtype([(('a', encodable_name), int)]).names[0], b'b')
+    @pytest.mark.skipif(sys.version_info[0] >= 3, reason="Not Python 2")
+    def test_unicode_field_names(self):
+        # Unicode field names are converted to ascii on Python 2:
+        encodable_name = u'b'
+        assert_equal(np.dtype([(encodable_name, int)]).names[0], b'b')
+        assert_equal(np.dtype([(('a', encodable_name), int)]).names[0], b'b')
 
-            # But raises UnicodeEncodeError if it can't be encoded:
-            nonencodable_name = u'\uc3bc'
-            assert_raises(UnicodeEncodeError, np.dtype, [(nonencodable_name, int)])
-            assert_raises(UnicodeEncodeError, np.dtype, [(('a', nonencodable_name), int)])
+        # But raises UnicodeEncodeError if it can't be encoded:
+        nonencodable_name = u'\uc3bc'
+        assert_raises(UnicodeEncodeError, np.dtype, [(nonencodable_name, int)])
+        assert_raises(UnicodeEncodeError, np.dtype, [(('a', nonencodable_name), int)])
 
-            # A single name string provided to fromarrays() is also allowed to be unicode:
-            x = np.core.records.fromarrays([[0], [0]], names=u'a,b', formats=u'i4,i4')
+        # A single name string provided to fromarrays() is also allowed to be unicode:
+        x = np.core.records.fromarrays([[0], [1]], names=u'a,b', formats=u'i4,i4')
+        assert_equal(x['a'][0], 0)
+        assert_equal(x['b'][0], 1)
 
-        def test_unicode_order(self):
-            # Test that we can sort with order as a unicode field name:
-            name = u'b'
-            x = np.zeros(1, dtype=[(name, int)])
-            x.sort(order=name)
+    @pytest.mark.skipif(sys.version_info[0] >= 3, reason="Not Python 2")
+    def test_unicode_order(self):
+        # Test that we can sort with order as a unicode field name:
+        name = u'b'
+        x = np.zeros(1, dtype=[(name, int)])
+        x.sort(order=name)
 
     def test_field_names(self):
         # Test unicode and 8-bit / byte strings can be used

